### ISSUE TITLE
silhouette score

### ISSUE THREAD

**Issue by kazemSafari** [facebookresearch/faiss#1875]
# Summary

Can you please provide support for calculating silhouette score on the gpu?
I can only find implementation of it using sklearn.metrics.silhouette_score and rapids.cuml at the moment.
Thank you.

OS: Ubuntu 20.04

Faiss version: 1.7.0

Installed from: anaconda

Faiss compilation options: IDK
Running on:
- GPU

Interface: 
- Python

**Comment by mdouze** [comment]
What is the definition of the silhouette score?

**Comment by kazemSafari** [comment]
You can find it here:
https://en.wikipedia.org/wiki/Silhouette_(clustering)

Also the sklearn implementation can be found here:
https://github.com/scikit-learn/scikit-learn/blob/15a949460/sklearn/metrics/cluster/_unsupervised.py#L38

It requires calculating the pairwise distances (based on a metric) between all points in the dataset.
I was wondering if you can implement it in a clever way so that it is not memory intensive for large datasets.

**Comment by sainivedh** [comment]
@kazemSafari Hi, have you found any GPU based Sihouette score implementations ?. 
cuml has it but facing cuda errors.

**Comment by KIC** [comment]
From my understanding you should be able to calculate this simply by:

```
def silhouette_coefficient(cluster_index, samples) -> float:
    # At each data point, we calculate the distance to the cluster center in which the data point belongs (referred to as a),
    # as well as the distance to the second best cluster center (referred to as b). Here, the second best cluster refers
    # to the closest cluster that is not the current data point’s cluster. Then based, on these two distances a and b, the
    # silhouette s of that data point is calculated as s=(b-a)/max(a,b). The mean of all s is the s coefficient and measures
    # quality of the clustering for this nr of k
    #
    # This is done by simply calculating the silhouette coefficient over a range of k, and identifying the peak as the optimum K.
    distance, _ = cluster_index.search(samples, 2)

    # a = distance[:, 0], b = distance[:, 1]
    # s = (b - a) / np.max(a, b)
    s = (distance[:, 1] - distance[:, 0]) / np.max(distance, 1)

    # A score of 1 denotes the best, meaning that the data point i is very compact within the cluster to which it belongs
    # and far away from the other clusters. The worst value is -1. Values near 0 denote overlapping clusters.
    return s.mean()
```
